src/plugins/discover/public/application/main/hooks/use_esql_mode.ts

```diff
@@ -57,55 +58,60 @@ export function useEsqlMode({
           if (!query || next.fetchStatus === FetchStatus.ERROR) {
             return;
           }
+
           const sendComplete = () => {
             stateContainer.dataState.data$.documents$.next({
               ...next,
               fetchStatus: FetchStatus.COMPLETE,
             });
           };
+
           const { viewMode } = stateContainer.appState.getState();
-          let nextColumns: string[] = [];
           const isEsqlQuery = isOfAggregateQueryType(query);
-          const hasResults = Boolean(next.result?.length);
-          let queryHasTransformationalCommands = false;
-          if ('esql' in query) {
-            if (hasTransformationalCommand(query.esql)) {
-              queryHasTransformationalCommands = true;
-            }
-          }
 
           if (isEsqlQuery) {
-            const language = getAggregateQueryMode(query);
+            const hasResults = Boolean(next.result?.length);
+
             if (next.fetchStatus !== FetchStatus.PARTIAL) {
               return;
             }
 
+            let nextColumns: string[] = prev.current.recentlyUpdatedToColumns;
+
             if (hasResults) {
-              // check if state needs to contain column transformation due to a different columns in the resultset
               const firstRow = next.result![0];
-              const firstRowColumns = Object.keys(firstRow.raw).slice(0, MAX_NUM_OF_COLUMNS);
-              if (!queryHasTransformationalCommands) {
-                nextColumns = [];
-                initialFetch.current = false;
+              const firstRowColumns = Object.keys(firstRow.raw);
+
+              if (hasTransformationalCommand(query.esql)) {
+                nextColumns = firstRowColumns.slice(0, MAX_NUM_OF_COLUMNS);
               } else {
-                nextColumns = firstRowColumns;
-                if (initialFetch.current && !prev.current.columns.length) {
-                  prev.current.columns = firstRowColumns;
-                }
+                nextColumns = [];
               }
             }
-            const addColumnsToState = !isEqual(nextColumns, prev.current.columns);
-            const queryChanged = query[language] !== prev.current.query;
+
+            if (initialFetch.current) {
+              initialFetch.current = false;
+              prev.current.query = query.esql;
+              prev.current.recentlyUpdatedToColumns = nextColumns;
+            }
+
+            const indexPatternChanged =
+              getIndexPatternFromESQLQuery(query.esql) !==
+              getIndexPatternFromESQLQuery(prev.current.query);
+
+            const addColumnsToState =
+              indexPatternChanged || !isEqual(nextColumns, prev.current.recentlyUpdatedToColumns);
+
             const changeViewMode = viewMode !== getValidViewMode({ viewMode, isEsqlMode: true });
-            if (!queryChanged || (!addColumnsToState && !changeViewMode)) {
+
+            if (!indexPatternChanged && !addColumnsToState && !changeViewMode) {
```

Comment:
QQ: if `indexPatternChanged` would be true, also `addColumnsToState`  would be true, so it would also work to go with 

```suggestion
            if (!addColumnsToState && !changeViewMode) {
```

Reply:
True although I would like to keep the current expression for readability.

Reply:
Yes I agree with Julia 👍 More readable this way.